internal/dynamic/plugin/input.go

```diff
@@ -0,0 +1,212 @@
+	}
+	value, err := runtimepb.AnyToProto(cfgValue)
+	if err != nil {
+		return false, fmt.Errorf("unable to convert config to proto: %w", err)
+	}
+	autoRetryNacks, err = backoff.RetryWithData(func() (bool, error) {
+		resp, err := client.Init(ctx, &runtimepb.BatchInputInitRequest{
+			Config: value,
+		})
+		if err != nil {
+			if !proc.IsRunning() {
+				return false, backoff.Permanent(fmt.Errorf("plugin exited early: %w", err))
+			}
+			return false, err
+		}
+		if err = runtimepb.ProtoToError(resp.Error); err != nil {
+			return false, backoff.Permanent(err)
+		}
+		return resp.AutoReplayNacks, nil
+	}, backoff.NewExponentialBackOff(backoff.WithMaxElapsedTime(maxStartupTime)))
+	if err != nil {
+		return false, fmt.Errorf("unable to initialize plugin: %w", err)
+	}
+	return autoRetryNacks, nil
+}
+
+// Connect implements service.BatchInput.
+func (i *input) Connect(ctx context.Context) (err error) {
+	var resp *runtimepb.BatchInputConnectResponse
+	for range retryCount {
+		resp, err = i.client.Connect(ctx, &runtimepb.BatchInputConnectRequest{})
+		if err != nil {
+			err = fmt.Errorf("unable to reach plugin: %w", err)
+			if i.proc.IsRunning() {
+				return err
+			}
+			if err := i.proc.Close(ctx); err != nil {
+				return fmt.Errorf("unable to retart plugin process: %w", err)
+			}
+			if _, err := startInputPlugin(ctx, i.proc, i.client, i.cfgValue); err != nil {
+				return fmt.Errorf("unable to restart plugin: %w", err)
+			}
+			continue
+		}
+		return nil
+	}
+	if err != nil {
+		return fmt.Errorf("unable to connect to plugin: %w", err)
+	}
+	return runtimepb.ProtoToError(resp.Error)
+}
+
+// ReadBatch implements service.BatchInput.
+func (i *input) ReadBatch(ctx context.Context) (service.MessageBatch, service.AckFunc, error) {
+	resp, err := i.client.ReadBatch(ctx, &runtimepb.BatchInputReadRequest{})
+	if err != nil {
+		if !i.proc.IsRunning() {
+			return nil, nil, service.ErrNotConnected
+		}
+		return nil, nil, fmt.Errorf("unable to read from plugin: %w", err)
+	}
+	if err := runtimepb.ProtoToError(resp.Error); err != nil {
+		return nil, nil, err
+	}
+	id := resp.BatchId
+	batch, err := runtimepb.ProtoToMessageBatch(resp.Batch)
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to convert batch from proto: %w", err)
+	}
+	return batch, func(ctx context.Context, err error) error {
+		resp, err := i.client.Ack(ctx, &runtimepb.BatchInputAckRequest{
+			BatchId: id,
+			Error:   runtimepb.ErrorToProto(err),
+		})
+		if err != nil {
+			return fmt.Errorf("unable to ack batch: %w", err)
+		}
+		return runtimepb.ProtoToError(resp.Error)
+	}, nil
+}
+
+// Close implements service.BatchInput.
+func (i *input) Close(ctx context.Context) error {
+	resp, err := i.client.Close(ctx, &runtimepb.BatchInputCloseRequest{})
+	if err != nil {
+		return fmt.Errorf("unable to close plugin: %w", err)
+	}
+	if err := runtimepb.ProtoToError(resp.Error); err != nil {
+		return fmt.Errorf("plugin close error: %w", err)
+	}
+	return i.proc.Close(ctx)
```

Comment:
How about wrapping the error for consistency?

Reply:
done